Adding agent internal telemetry

iglendd opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?
Adds ability to send and query cross-org telemetry on a few internal Agent metrics. Specifically
* Sending a few internal and built-in checks.*, logs.*, transactions.*, dogstatsd.* metrics
* Disabled by default.
* Default schedule sent 30 seconds after agent start and every 15 minutes after that

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes
<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->
* Agent check telemetry (telemetry.check) is activated when agent_telemetry.enabled, but resulting metrics will not be send automatically.
* We need to makes sure that agent memory usage will not grow and RSS remains stables (e.g. in cases if large number of metrics is collected and thrown away from Gather() method.

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes
* Set agent_telemetry.enabled: true
* Restart agent
* Wait a few minutes minute
* Go to datadog org UI
* Select Log Explorer menu
* Paste in a search bar `source:agent-telemetry` - you should see a handful of messages with `Agent status` and `Agent metric` content, which contain JSON payloads.

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

iglendd committed:
First cut of internal Agent Telemetry

* Sending a few internal and built-in checks.*, logs.*, transactions.*, dogstatsd.* metrics and statuses
* Disabled by default
* Default schedule sent 30 seconds after agent start and every 15 minutes after that
* Sent to the org's instrumentation-telemetry-intake

iglendd committed:
Experimental, moving out agenttelemetry component from comp/core bundle

buraizu reviewed (dismissed):
Thanks for the PR, I've left some minor updates to the release note

vickenty commented on comp/core/agenttelemetry/agenttelemetryimpl/config.go:
```diff
@@ -0,0 +1,418 @@
+        - name: logs.network_errors
+        - name: logs.dropped
+        - name: logs.sender_latency
+        - name: logs.destination_http_resp
```
This metric has `url` tag that may leak sensitive info. Is that expected?

Since all these metrics are defined elsewhere, it would be easy to add tags with sensitive information by mistake. Would it make sense to allow-list all tags?

iglendd replied:
>  This metric has url tag that may leak sensitive info. Is that expected?

Not sure about danger of url in terms of privacy but honestly I do not know what urls can be appeared there. I presumened it was Datadog domain url. 

> Since all these metrics are defined elsewhere, it would be easy to add tags with sensitive information by mistake. Would it make sense to allow-list all tags?
True, it did concerned me. It certainly has big hole when privacy may be inadvertently violated  by any developer without even realizing it. I will add facility to specify explicit tag management. I did not have time to do it before.

iglendd replied:
I believe that I have addressed concerns in sub-sequent commit by adding to configuration needs to explicitly specify associated tags. If not tags are specified, all of them will be dropped (but timeseries aggregated accordingly)

vickenty commented on comp/core/agenttelemetry/agenttelemetryimpl/config.go:
```diff
@@ -0,0 +1,418 @@
+	statusExtraBuilder []jBuilder
+	metricsMap         map[string]*MetricConfig
+	excludeZeroMetric  bool
+	excludeTags        map[string]any
```
It seems we only assign `struct{}{}` here, would this work?
```suggestion
	excludeTags        map[string]struct{}
```

iglendd replied:
> It seems we only assign struct{}{} here, would this work?

No, actually. Because excluded tag can be specified with or without value. And if there is a value I change the node to another map. See https://github.com/DataDog/datadog-agent/pull/22771/files#diff-4d02442e857d5db8627ad2e0fc6f7cccc9dfff7e0765db5448878d3cfc817935R282-R284. Before `any` I did have `interface{}` type before.

vickenty commented on pkg/status/render/templates/agent-telemetry-basic.tmpl:
```diff
@@ -0,0 +1,18 @@
+  },
+  {{- end }}
+  {{- with .DogStatsdStats -}}
+  {{- $alreadyIterated := 0 }}
```
Would it be possible to create this json payload with `encoding/json` instead?

Also, would it make sense to list specific data items to include, rather than doing it all in bulk? So that future modifications to dogstatsd or checks do not accidentally expose sensitive data (which otherwise is fine in agent status output or inventories).

iglendd replied:
> Would it be possible to create this json payload with encoding/json instead?

Can you please clarify what do you mean exactly. Specifically what to add/change. The template is rendered as a []byte and later umarlshalle by encodiog/json (https://github.com/DataDog/datadog-agent/pull/22771/files#diff-1801173c4f68bdc061d17c1a22740783fd535108faafeae5c5d63e02b049312eR264)

> Also, would it make sense to list specific data items to include, rather than doing it all in bulk? So that future modifications to dogstatsd or checks do not accidentally expose sensitive data (which otherwise is fine in agent status output or inventories).

Yes, it would I make a necessary adjustments. Thank you.

iglendd replied:
Ah, my main goal was to avoid hardcoding transformation processing and simplify its management by changing mappings in the template. Especially if I need a different types of transformations. The amount of input/output, and frequency of its execution makes efficiency question moot IMHO but certainly help flexibility of transformation modification. And it appears that what templates are used across of agent as well and some of them to generate JSON from JSON, no?

iglendd replied:
I did not change that logic (and was not intended) but I have made a few improvements in last commit since (per discussion with @GustavoCaso). I have moved rendering and template itself into agenttelemetry component. And I have requesting now only sections are used during rendering and no more.

iglendd replied:
We can always discuss it further if needed as I did with @GustavoCaso. I am closing it for now, please notify me if it is not acceptable at this time.

iglendd commented on pkg/config/setup/config.go:
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
+	// Agent Telemetry
+	// Enable various telemetry on the internals of the Agent.
+	// This will not tirgger any billable custom metrics.
+	config.BindEnvAndSetDefault("agent_telemetry.enabled", false)
```
I do not have strong preference. When I started spend some time thinking of reusing telemetry name space or package but in the end decided against it due to drastically different semantics and desire to avoid breaking existing code and behavior. It is also a different code hierarchy now. I do hope that at some point I would be able to to merge either agent_telemetry into telemetry package/component or other way around.

iglendd replied:
I am closing for now. We can always revisit it later.

iglendd committed:
Address code review comment by referencing telemetry sources explicitly

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:
```diff
@@ -0,0 +1,537 @@
+	cfgComp    config.Component
+	logComp    log.Component
+	telComp    telemetry.Component
+	statusComp status.Component
```
I'm confused about the need to use the status component as part of this new component.

Usually how things operate for the status component is that it aggregates information from other components via [FX value groups](https://uber-go.github.io/fx/value-groups/) and then when the customer calls `agent status` we display the information of each component.

It feels a bit weird how this component is using the status component. Ex [logs agent status provider](https://github.com/DataDog/datadog-agent/blob/2a41ad20c3f5051f3f0b9db2e0c23bda630a5bd0/comp/logs/agent/agent.go#L71). In the example the logs agent defines a status provider which would be register by the status component and later render. 

@iglendd I would like to know more about why you decided to use the status component in this way.

Also, just to be clear there is a lack of documentation on how to register status providers. I'm currently in the middle of the migration, so I will write a document explaining how to do it shortly.

iglendd replied:
My intent was extract parts of status JSON which is reach source of Agent Telemetry, I am using template and JQ for that. I am not providing or contributing to the status (yet).

iglendd replied:
I have addressed some of the concerns. Please review the last commit which reflects our discussion.

GustavoCaso reviewed (changes requested):
@iglendd I would love to have chat about the use of the status component to make sure we are levering the status component correctly.

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/config.go:
```diff
@@ -0,0 +1,526 @@
+	if len(atCfgMap) > 0 {
+		// Reconvert to string and back to object.
+		// Config.UnmarshalKey() is better but it did not work in some cases
+		atCfgBytes, err := yaml.Marshal(atCfgMap)
```
I'm not entire sure how are config parsing logic works 100%. Does the first pass to the config parsed nested configuration? Mostly asking to make sure we avoid doing any extra `yaml.Marshal`

iglendd replied:
I have hard time to extract more complex configuration hierarchy. One method work but it seems to have limitations. My double marshal/unmarshall is a work around for that since I do want to have unmarshal into dedicated structure. I did left a comment of that sort.

iglendd replied:
Closing per OOTB discussion

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/utils.go:
```diff
@@ -0,0 +1,161 @@
+)
+
+type jBuilder struct {
+	jqSource    *gojq.Query
```
lThe idea of running JQ directly sounds good. It has a penalty on our side that we need to bring a 3rd party dependency into the mix. 

Is there any other way to achieve the same logic without adding support for JQ?

iglendd replied:
We already use this module (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/jsonquery/jsonquery.go#L13), which is further indirectly used in https://github.com/DataDog/datadog-agent/blob/main/pkg/compliance/resolver.go#L384 and https://github.com/DataDog/datadog-agent/blob/main/cmd/agent/common/autodiscovery.go#L319. Note to myself. Switch to our own wrapper - jsonquery

GustavoCaso replied:
Thanks for the context.

If I understand we use JQ to be able to modify the metrcis we emit juts by modifying the configuration, which would mean that we are able to change things faster rather than deploying the agent.

I still think that having this feature been use by customer (Still experimental) and being this dynamic might lead to unwanted consequences. Maybe a customer runs a complex JQ query (performance) Or creates an Invalid JQ  (shoot themself in the foot). Hard to migrate from using JQ (Once customers use it it might be difficult to remove t)

I'm raising this concerns, because we might reconsider shipping the initial implementation without JQ and see how adoption is and validate if we need this functionality. 

For now we could remove the `extra` feature:
```
extra:
        runner.workers.count: '.runnerStats.Workers.Count'
        runner.workers.utilization: '[.runnerStats.Workers | .. | objects | .Utilization] | add'
```

Extract those values manually for the `basic` template.

iglendd replied:
discussed in OOTB

GustavoCaso commented on pkg/config/utils/telemetry.go:
```diff
@@ -32,5 +37,11 @@ func IsCheckTelemetryEnabled(checkName string, cfg pkgconfigmodel.Reader) bool {
+}
+
+// IsAgentTelemetryEnabled returns whether or not agent telemetry is enabled
+func IsAgentTelemetryEnabled(cfg pkgconfigmodel.Reader) bool {
```
Every time we make a public function on this package tends to be use in many places 😢 Ideally in the future we would have a single configuration component 🎉 

Since this function is only use here could we make it private or inline it?

iglendd replied:
Right, made it private. Ideally we would like to join these two configurations in the future someway. @vickenty suggested to use a slightly different naming to indicate similar lineage (see his comment above)

olivielpeau commented on pkg/config/setup/config.go:
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
 	// The histogram buckets use to track the time in nanoseconds it takes for a DogStatsD listeners to push data to the server
 	config.BindEnvAndSetDefault("telemetry.dogstatsd.listeners_channel_latency_buckets", []string{})
 
+	// Agent Telemetry
```
Please mention here that the feature is experimental and should not be enabled unless prompted by Datadog Support

iglendd replied:
Agree, fixed in the last commit

olivielpeau commented on pkg/config/setup/config.go:
```diff
@@ -1086,6 +1086,11 @@ func InitConfig(config pkgconfigmodel.Config) {
 
+	// Agent Telemetry
+	// Enable various telemetry on the internals of the Agent.
+	// This will not tirgger any billable custom metrics.
```
Let's remove this mention for now, can be re-introduced once the feature is not experimental anymore

iglendd replied:
Agree, fixed in the last commit

iglendd committed:
Code review related changes

* Moved status rendering into agenttelemetry component
* Retrieved only used status provider
* Few other cosmetic changes

iglendd commented on comp/core/agenttelemetry/agenttelemetryimpl/config.go:
```diff
@@ -0,0 +1,526 @@
+
+	// Compile status extra
+	var err error
+	p.statusExtraBuilder, err = compileJBuilders(p.Status.Extra)
```
To me it looks like it would muddy config.go file content a little bit but otherwise it is not principal.

iglendd committed:
Lint fixes

iglendd commented:
@GustavoCaso and @vickenty please re-review when you have a chance. I am not planning to merge before Privacy review is completed but I need to run tests on staging. I would like to rebase ideally soon but afraid it will screw up reviews and its positions.

GustavoCaso commented on comp/core/status/statusimpl/status_test.go:
```diff
@@ -927,7 +927,7 @@ Status render errors
 	for _, testCase := range testCases {
 		t.Run(testCase.name, func(t *testing.T) {
-			bytesResult, err := statusComponent.GetStatusBySection(testCase.section, testCase.format, false)
+			bytesResult, err := statusComponent.GetStatusBySection([]string{testCase.section}, testCase.format, false)
```
Could we add a test case that uses multiple sections?
Another possible test is to make sure if the section do not exists or is empty dopes not cause any issue.

iglendd replied:
Done in last commit

GustavoCaso commented on comp/core/status/component.go:
```diff
@@ -21,8 +21,8 @@ const CollectorSection string = "collector"
-	// Returns only the agent status for the especify section and format type
-	GetStatusBySection(section string, format string, verbose bool) ([]byte, error)
+	// Returns only the agent status for the specified sections and format type
+	GetStatusBySection(sections []string, format string, verbose bool) ([]byte, error)
```
Should we rename the function to `GetStatusBySections` ?

iglendd replied:
Sure, that is easy

iglendd replied:
see it in last commit

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:
```diff
@@ -0,0 +1,548 @@
+
+	templateName := "agent-telemetry-" + p.Status.Template + ".tmpl"
+	var b = new(bytes.Buffer)
+	err := status.RenderText(templatesFS, templateName, b, inputStatus)
```
🎉

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry.go:
```diff
@@ -0,0 +1,548 @@
+
+	a.logComp.Debugf("Collect Agent Status telemetry for profile %s", p.Name)
+
+	// Current "agent-telemetry-basic.tmpl" uses only "runneStats" and "dogstatsdStats" JSON sections
+	// These JSON sections are populated via "collector" and "DogStatsD" status providers sections
+	minimumReqSections := []string{"collector", "DogStatsD"}
```
nit: Since we might have more templates in the future with different required sections. We could use a map to store template and status sections information

iglendd replied:
True, but I want to see it first.

GustavoCaso commented on comp/core/agenttelemetry/agenttelemetryimpl/agenttelemetry_test.go:
```diff
@@ -0,0 +1,495 @@
+	return hostMock{}
+}
+
+// Status component currently has mock but it appears to be not compatible with fx  fx fails
+type statusMock struct {
+}
+
+func (s statusMock) GetStatus(string, bool, ...string) ([]byte, error) {
+	return []byte{}, nil
+}
+func (s statusMock) GetStatusBySection([]string, string, bool) ([]byte, error) {
+	return []byte{}, nil
+}
+func newStatusMock() statusMock {
+	return statusMock{}
+}
```
I'll add it to my TODO list 😄

iglendd committed:
Code review related changes

iglendd commented:
@GustavoCaso I have addressed most of your review comments (please check it out). @vickenty can you finish your review please. I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.

vickenty commented:
> I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.

I think merging works a bit better than rebasing once review has started, so if possible and practical, please merge.

iglendd commented:
> > I also have a question, the PR is falling behind. Do you prefer I rebase or merge main in (there is one conflict to resolve too). Thank you.
> 
> I think merging works a bit better than rebasing once review has started, so if possible and practical, please merge.

Of course. I will do merging, thank you.

GustavoCaso commented:
@iglendd I would like to draw some attention to my previous comment https://github.com/DataDog/datadog-agent/pull/22771#discussion_r1514260175

I just realized one thing about allowing arbitrary JQ functionality to run against the status JSON output. Once customers start to use that functionality we won't be able to modify the status JSON output without fear of breaking any telemetry `extra` custom metrics. 

I really want to surface that to see if we could consider my suggestion of shipping the first experimental version with support for `extra` 

I'm more than happy to chat about this via Slack, Zoom or whatever works best for you

iglendd committed:
Code review related changes

iglendd committed:
Merge branch 'main' into len.gamburg/agent-telemetry

iglendd committed:
Improve code after merge

iglendd committed:
Limit support only to site instead of dd_url configuration

iglendd committed:
Merge branch 'main' into len.gamburg/agent-telemetry

iglendd committed:
Fixed affected by last change unit tests

iglendd commented:
> @iglendd I would like to draw some attention to my previous comment [#22771 (comment)](https://github.com/DataDog/datadog-agent/pull/22771#discussion_r1514260175)
> 
> I just realized one thing about allowing arbitrary JQ functionality to run against the status JSON output. Once customers start to use that functionality we won't be able to modify the status JSON output without fear of breaking any telemetry `extra` custom metrics.
> 
> I really want to surface that to see if we could consider my suggestion of shipping the first experimental version with support for `extra`
> 
> I'm more than happy to chat about this via Slack, Zoom or whatever works best for you

Chatted OOTB

iglendd committed:
Fix a few linters problems

iglendd committed:
Merge branch 'main' into len.gamburg/agent-telemetry

iglendd committed:
Fix results of a faulty linter on Windows

iglendd committed:
Fix another linter related problem

buraizu commented on comp/README.md:
```diff
@@ -69,6 +69,12 @@ Package collector defines the collector component.
+
+*Datadog Team*: agent-metrics-logs
+
+Package agenttelemetry implements a component to generate agent telemetry
```
```suggestion
Package agenttelemetry implements a component to generate Agent telemetry
```

iglendd replied:
Thank you @buraizu.  The change is accepted.

buraizu reviewed (approved):
Thanks, approving with a minor update requested

GustavoCaso commented on comp/metadata/host/hostimpl/utils/host_windows.go:
```diff
@@ -38,29 +39,12 @@ var (
 )
 
-// InfoStat describes the host status.  This is not in the psutil but it useful.
-type InfoStat struct {
-	Hostname             string `json:"hostname"`
-	Uptime               uint64 `json:"uptime"`
-	BootTime             uint64 `json:"bootTime"`
-	Procs                uint64 `json:"procs"`           // number of processes
-	OS                   string `json:"os"`              // ex: freebsd, linux
-	Platform             string `json:"platform"`        // ex: ubuntu, linuxmint
-	PlatformFamily       string `json:"platformFamily"`  // ex: debian, rhel
-	PlatformVersion      string `json:"platformVersion"` // version of the complete OS
-	KernelVersion        string `json:"kernelVersion"`   // version of the OS kernel (if available)
-	KernelArch           string `json:"kernelArch"`
-	VirtualizationSystem string `json:"virtualizationSystem"`
-	VirtualizationRole   string `json:"virtualizationRole"` // guest or host
-	HostID               string `json:"hostid"`             // ex: uuid
-}
-
```
@iglendd is this change intentional? 

I'm going to flag to @hush-hush who wrote the host implementation to make sure we are ok with this change

hush-hush replied:
Looking through the history this was linked to no longer using WMI and removing usage of gopsutil from windows (#1725).

This code has been moved a number of times, we should definitely add a comment to explain this.

GustavoCaso reviewed (approved):
Len changes looks good. 

We have discussed extensible over Zoom and Slack. The use of the status component and JQ is to enable sending telemetry metrics as easy as possible. Knowing that feature is experimental and that is subject to change in the future, I would like for us to reconsider in the future the use of a similar pattern from the status component as each component registering there own telemetry metrics. Ensuring decoupling between components. Today the telemetry component has strong dependency on the status component. Adding new information for the telemetry agent means having to add it the status provider of a component. 
If we have that in mind and in future iterations we are able to remove that dependency I think we would be a in a better position to evolve the telemetry component in isolation. 

One more thing that would interesting to add in the PR description is amore detail explanation of the QA steps. ex. `Paste in a search bar source:agent-telemetry` Which search bar? Having as much detail information on how todo the QA the better. 

Len, thanks for the patience with all my comments and for taking the time to explain me the needs for the telemetry work.

iglendd committed:
Removed original host componenet enchancement per code review

iglendd committed:
Casing changes per code review

iglendd committed:
Restore original host implementation for windows since no generalized component is needed (per code review)

iglendd committed:
Added fake commit to address gofmt complains on next commit

iglendd committed:
Commit to address gofmt linter issues

iglendd committed:
Still fixing gofmt issues

iglendd committed:
Fixing gofmt issues

iglendd committed:
Merge branch 'main' into len.gamburg/agent-telemetry

iglendd commented:
> Knowing that feature is experimental and that is subject to change in the future, I would like for us to reconsider in the future the use of a similar pattern from the status component as each component registering there own telemetry metrics. Ensuring decoupling between components. Today the telemetry component has strong dependency on the status component. Adding new information for the telemetry agent means having to add it the status provider of a component. If we have that in mind and in future iterations we are able to remove that dependency I think we would be a in a better position to evolve the telemetry component in isolation.

Definitely. I also planning to start using our DefaultForwarder to replace portion of sender (cc @hush-hush) 

> One more thing that would interesting to add in the PR description is amore detail explanation of the QA steps. ex. `Paste in a search bar source:agent-telemetry` Which search bar? Having as much detail information on how todo the QA the better.

I have added a few more items in the step, I hope it is a bit better.

iglendd committed:
Merge branch 'main' into len.gamburg/agent-telemetry

iglendd committed:
Fix merge errors

iglendd committed:
Fix build problems

iglendd committed:
Fix build problems

iglendd committed:
Fix merge problem

iglendd committed:
Merge remote-tracking branch 'origin/main' into len.gamburg/agent-telemetry

iglendd committed:
Anothe build after merge problem

iglendd committed:
Anothe build after merge problem

iglendd committed:
Yet another fix build after merge

iglendd commented:
/merge

iglendd commented:
/merge
